### PR TITLE
2D Distance Driven

### PR DESCRIPTION
**Pull request by SoniaMinseoKim**

### CODE REVIEW

File: 2DIterativeRecon.jl [hunk 1]
```diff
@@ -0,0 +1,431 @@
...
+
+# Tube
+tubeX = 0     
+tubeY = DSO    
+
+# Iso-center
+isoX = geo.isoX
+isoY = geo.isoY
+
+sinogram = zeros(size(theta,1),nDet)
```
Comment (JeffFessler) [inline 2DIterativeRecon.jl:51]:
```suggestion
sinogram = zeros(T, nDet, length(theta))
```

eventually i prefer this way, after it all works...

File: 2DIterativeRecon.jl [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+# Geometry
+deg = 1
+geo = (DSD = 100000, DSO = 99700, pSize = 1, dSize = 0.5, nPix = 256, nDet = 1024
+        , theta = deg2rad.(0:deg:360-deg), isoX = 0, isoY = 0) #Define tuple
+
+# Make the Phantom Image
+phantomImg = shepp_logan(geo.nPix, SheppLoganToft())
+
+## Projection Branchless
+function projectionBranchless(phantom,geo ; draw::Bool = false)
```
Comment (JeffFessler) [inline 2DIterativeRecon.jl:24]:
```suggestion
function projectionBranchless(phantom::AbstractMatrix{<:T}, geo ; draw::Bool = false) where {T <: Number}
```

File: 2DIterativeRecon.jl [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+tubeY = DSO    
+
+# Iso-center
+isoX = geo.isoX
+isoY = geo.isoY
+
+sinogram = zeros(size(theta,1),nDet)
+
+# For each projection
+for proj=1:size(theta,1)
```
Comment (JeffFessler) [inline 2DIterativeRecon.jl:54]:
```suggestion
for proj=1:length(theta)
```

File: 2DIterativeRecon.jl [hunk 4]
```diff
@@ -0,0 +1,431 @@
...
+## Projection Branchless
+function projectionBranchless(phantom,geo ; draw::Bool = false)
+
+DSD = geo.DSD      
+DSO = geo.DSO      
+pSize = geo.pSize    
+dSize = geo.dSize    
+nPix = geo.nPix      
+nDet = geo.nDet     
+theta = geo.theta
```
Comment (JeffFessler) [inline 2DIterativeRecon.jl:32]:
```suggestion
theta = geo.theta # vector of projection view angles in radians
```

File: 2DIterativeRecon.jl [hunk 5]
```diff
@@ -0,0 +1,431 @@
...
+isoY = geo.isoY
+
+sinogram = zeros(size(theta,1),nDet)
+
+# For each projection
+for proj=1:size(theta,1)
+    
+    angle = theta[proj]
+       
+    # Tubre rotation
```
Comment (JeffFessler) [inline 2DIterativeRecon.jl:58]:
```suggestion
    # Tube rotation
```

File: 2DIterativeRecon.jl [hunk 6]
```diff
@@ -0,0 +1,431 @@
...
+
+    # Detector rotation
+    rdetX = ( (detX .- isoX ).*cos(angle) - (detY .- isoY ).*sin(angle) ) .+ isoX
+    rdetY = ( (detX .- isoX ).*sin(angle) + (detY .- isoY ).*cos(angle) ) .+ isoY
+    
+    if (draw == true)
+        drawGeo(rtubeX,rtubeY,rdetX,rdetY)   
+    end
+
+    # Define angle case & which axis it it project boundaries
```
Comment (JeffFessler) [inline 2DIterativeRecon.jl:70]:
```suggestion
    # Define angle case & which axis it project boundaries
```

File: 2DIterativeRecon.jl [hunk 7]
```diff
@@ -0,0 +1,431 @@
...
+            # Case 4
+                axisXCase = false   # Map on Y axis()
+                angleCase = 4
+                c1=0
+                c2=-1
+            end
+        end
+    end
+    
+    # Mapping boundaries into a commum axis()
```
Comment (JeffFessler) [inline 2DIterativeRecon.jl:101]:
```suggestion
    # Map boundaries into a common axis()
```

File: 2DIterativeRecon.jl [hunk 8]
```diff
@@ -0,0 +1,431 @@
...
+        end
+    else
+        # X-Ray pixel intersection calculation
+        L = abs(pSize/sin(angle)) # This account for parallel-beam
+        # Correction for fan-beam
+        for n=1:nDet
+            L1[n]=sqrt((rtubeX-0)^2+(rtubeY-detm[center_det])^2)/sqrt((rtubeX-0)^2+(rtubeY-detm[n])^2)
+        end           
+    end     
+    L = L./L1
```
Comment (JeffFessler) [inline 2DIterativeRecon.jl:131]:
```suggestion
    L = L ./ L1
```

File: 2DIterativeRecon.jl [hunk 9]
```diff
@@ -0,0 +1,431 @@
...
+    else
+        detIstart = nDet+1
+        detIinc = -1
+    end
+
+    
+    deltaDetm = detm[detIstart+detIinc] - detm[detIstart] # Mapped detector length()
+    deltaPixm = pixm[1,2] - pixm[1,1]   # Mapped pixel length()
+    
+    sinoTmp = zeros(1,nDet)
```
Comment (JeffFessler) [inline 2DIterativeRecon.jl:147]:
```suggestion
    sinoTmp = zeros(T, nDet) # one "row" of sinogram (column vector)
```

File: 2DIterativeRecon.jl [hunk 10]
```diff
@@ -0,0 +1,431 @@
...
+    
+    # For each row
```
Comment (JeffFessler) [inline 2DIterativeRecon.jl:149]:
```suggestion
    # For each image row
```

File: 2DIterativeRecon.jl [hunk 11]
```diff
@@ -0,0 +1,431 @@
...
+# Julia is compiling program so you need to restart the program in order to clear the variables 
+
+## Geometry Definitions
+# Geometry
+deg = 1
+geo = (DSD = 100000, DSO = 99700, pSize = 1, dSize = 0.5, nPix = 256, nDet = 1024
+        , theta = deg2rad.(0:deg:360-deg), isoX = 0, isoY = 0) #Define tuple
+
+# Make the Phantom Image
+phantomImg = shepp_logan(geo.nPix, SheppLoganToft())
```
Comment (JeffFessler) [inline 2DIterativeRecon.jl:21]: this test data should go at *bottom* of file (and eventually in a separate test file)

File: 2DIterativeRecon.jl [hunk 12]
```diff
@@ -0,0 +1,431 @@
...
+end
+
+## Draw 2D
+# Function to draw geometry
+function drawGeo(tubeX,tubeY,detX,detY)
+    plot(tubeX,tubeY,markershape = :asterisk)
+    plot!([detX[1],detX[end]],[detY[1],detY[end]])
+end
+
+
```
Comment (JeffFessler) [inline 2DIterativeRecon.jl:410]:
```suggestion
# Begin tests
```

**Review by JeffFessler — CHANGES_REQUESTED** [review]
I made several code suggestions for you Sonia.

File: src/dd2.jl [hunk 13]
```diff
@@ -1,20 +1,17 @@
 ## Matlab Author: Rodrigo de Barros Vimieiro
 ## Translated by Sonia Minseo Kim
 
-## 2-D Branchless Distance Driven Code
```
Comment (JeffFessler) [inline src/dd2.jl]: Why did you remove "Branchless" here?
There are two versions of DD: regular and branchless.
I think this code is the branchless version so we should say that, right?

Comment (SoniaMinseoKim) [reply src/dd2.jl]: I thought it was technically not branchless because we have the if statements

Comment (JeffFessler) [reply src/dd2.jl]: The inner-most loop has no `if` statements, and that is what matters for performance!

Comment (SoniaMinseoKim) [reply src/dd2.jl]: I see! And on a side note, this version of the code works now. Is there anything else you think I need to change?

File: src/bdd_2d.jl [hunk 14]
```diff
@@ -0,0 +1,381 @@
+## Matlab Author: Rodrigo de Barros Vimieiro
+## Translated by Sonia Minseo Kim
+
+## 2-D Branchless Distance Driven Code
+using Plots: plot
```
Comment (JeffFessler) [inline src/bdd_2d.jl:5]:
```suggestion
#using Plots: plot
```

Plots is an expensive dependency so we want to avoid it.  We can just comment it out.

File: src/bdd_2d.jl [hunk 15]
```diff
@@ -0,0 +1,381 @@
...
+        # Tube rotation
+        rtubeX = ( (tubeX - isoX )*cos(angle) - (tubeY - isoY )*sin(angle) ) + isoX
+        rtubeY = ( (tubeX - isoX )*sin(angle) + (tubeY - isoY )*cos(angle) ) + isoY
+
+        # Detector rotation
+        rdetX = ( (detX .- isoX ).*cos(angle) - (detY .- isoY ).*sin(angle) ) .+ isoX
+        rdetY = ( (detX .- isoX ).*sin(angle) + (detY .- isoY ).*cos(angle) ) .+ isoY
+        
+        if (draw == true)
+            drawGeo(rtubeX,rtubeY,rdetX,rdetY)
```
Comment (JeffFessler) [inline src/bdd_2d.jl:54]:
```suggestion
#           drawGeo(rtubeX,rtubeY,rdetX,rdetY)   
```

File: src/bdd_2d.jl [hunk 16]
```diff
@@ -0,0 +1,381 @@
...
+    return y
+end
+
+## Map X
+# Function that map detector | pixel boundaries onto X axis()
+function mapp2x(x1,y1,x2,y2)
+    x = -(y1)*(x1-x2)/(y1-y2)+x1
+    return x
+end
+
```
Comment (JeffFessler) [inline src/bdd_2d.jl:375]:
```suggestion
#=
```

File: src/bdd_2d.jl [hunk 17]
```diff
@@ -0,0 +1,381 @@
...
+## Draw 2D
+# Function to draw geometry
+function drawGeo(tubeX,tubeY,detX,detY)
+    plot(tubeX,tubeY,markershape = :asterisk)
+    plot!([detX[1],detX[end]],[detY[1],detY[end]])
+end
```
Comment (JeffFessler) [inline src/bdd_2d.jl:381]:
```suggestion
end
=#
```
It was good for testing, but not needed in final version...

File: test/bdd_2d.jl [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+@testset "bdd_2d" begin
+    deg = 1
+    geo = (DSD = 100000, DSO = 99700, pSize = 1, dSize = 0.5, nPix = 256, nDet = 1024,
+          theta = deg2rad.(0:deg:360-deg), isoX = 0, isoY = 0)
+    phantomImg = shepp_logan(geo.nPix, SheppLoganToft())
+    sinogramB = projection(phantomImg',geo)
+    imageB = backprojection(sinogramB,geo)
+
+    sinogramB_correct = load("sinomat.jld")["sinomat"]
+    imageB_correct = load("imagemat.jld")["imagemat"]
```
Comment (JeffFessler) [inline test/bdd_2d.jl:14]:
```suggestion
#   sinogramB_correct = load("sinomat.jld")["sinomat"]
#   imageB_correct = load("imagemat.jld")["imagemat"]
```
A code repo should only have source code in it, no data files like .jld files.
It was great that you tested it this way, but please remove the .jld files from your PR.

File: test/bdd_2d.jl [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+
+    sino = @inferred projection(phantomImg',geo)
+    @test sino isa Matrix
+    @test size(sinogramB) == (360, 1024)
+    image = @inferred backprojection(sinogramB,geo)
+    @test image isa Matrix
+    @test size(imageB) == (256, 256)
+
+    @test sinogramB == sinogramB_correct
+    @test imageB == imageB_correct
```
Comment (JeffFessler) [inline test/bdd_2d.jl:24]:
```suggestion
#   @test sinogramB == sinogramB_correct
#   @test imageB == imageB_correct
```
Did this test actually run for you?  It would be rare that `==` would work for floating point data!
Let's discuss how to check the tests when we meet next.

File: src/bdd_2d.jl [hunk 20]
```diff
@@ -0,0 +1,381 @@
+## Matlab Author: Rodrigo de Barros Vimieiro
+## Translated by Sonia Minseo Kim
+
+## 2-D Branchless Distance Driven Code
+using Plots: plot
+using ImagePhantoms: shepp_logan, SheppLoganToft
+using LazyGrids: ndgrid 
+using Dierckx: Spline1D
+
+## Projection
```
Comment (JeffFessler) [inline src/bdd_2d.jl:10]:
```suggestion
## Projection
"""
     projection(phantom::AbstractMatrix{<:T}, geo)
(you need to document the usage here in a docstring, for every exported function.  see other examples)
"""
```

File: src/bdd_2d.jl [hunk 21]
```diff
@@ -0,0 +1,381 @@
+## Matlab Author: Rodrigo de Barros Vimieiro
+## Translated by Sonia Minseo Kim
+
+## 2-D Branchless Distance Driven Code
+using Plots: plot
+using ImagePhantoms: shepp_logan, SheppLoganToft
+using LazyGrids: ndgrid 
+using Dierckx: Spline1D
+
```
Comment (JeffFessler) [inline src/bdd_2d.jl:9]: Need `export` statements near here - see other code for examples.

**Review by JeffFessler — CHANGES_REQUESTED** [review]

File: src/bdd_2d.jl [hunk 22]
```diff
@@ -0,0 +1,410 @@
...
+
+
+## Integrate function
+"""
+    integrate1D(p_v::Vector, pixelSize)
+Calculates the integral image set,
+for a given column vector and a pixel size 
+"""
+function integrate1D(p_v::Vector,pixelSize)
```
Comment (JeffFessler) [inline src/bdd_2d.jl:362]:
```suggestion
function integrate1D(p_v::Vector{T}, pixelSize) where {T <: Number}
```

File: src/bdd_2d.jl [hunk 23]
```diff
@@ -0,0 +1,410 @@
...
+
+    n_pixel = length(p_v)
+    
+    P_x = 0
+    Ppj = zeros(n_pixel+1)
```
Comment (JeffFessler) [inline src/bdd_2d.jl:367]:
```suggestion
    Ppj = zeros(T, n_pixel+1)
```

File: src/bdd_2d.jl [hunk 24]
```diff
@@ -0,0 +1,410 @@
+## Matlab Author: Rodrigo de Barros Vimieiro
+## Translated by Sonia Minseo Kim in 2022
+
+## 2-D Branchless Distance Driven Code
+
+export bdd_2d
+
+# using Plots: plot
+using ImagePhantoms: shepp_logan, SheppLoganToft
```
Comment (JeffFessler) [inline src/bdd_2d.jl:9]:
```suggestion
```
These are for tests only - not for the src code.

File: test/runtests.jl [hunk 25]
```diff
@@ -6,7 +6,7 @@ using Sinograms
 @testset "Sinograms" begin
     # todo
```
Comment (JeffFessler) [inline test/runtests.jl:7]:
```suggestion
```

File: src/bdd_2d.jl [hunk 26]
```diff
@@ -0,0 +1,408 @@
...
+
+
+"""
+    mapp2x(x1,y1,x2,y2)
+Maps detector or pixel boundaries onto x-axis,
+for tube and detector rotation angles and detector/pixel boundaries
+"""
+function mapp2x(x1,y1,x2,y2)
+    x = -(y1)*(x1-x2)/(y1-y2)+x1
+    return x
```
Comment (JeffFessler) [inline src/bdd_2d.jl:398]:
```suggestion
    return -y1*(x1-x2)/(y1-y2) + x1
```

File: src/bdd_2d.jl [hunk 27]
```diff
@@ -0,0 +1,408 @@
...
+
+    # Tube
+    tubeX = 0     
+    tubeY = DSO    
+
+    # Iso-center
+    isoX = geo.isoX
+    isoY = geo.isoY
+
+    sinogram = zeros(length(theta),nDet)
```
Comment (JeffFessler) [inline src/bdd_2d.jl:44]:
```suggestion
    sinogram = zeros(T, length(theta), nDet)
```

File: src/bdd_2d.jl [hunk 28]
```diff
@@ -0,0 +1,408 @@
...
+
+    # Tube
+    tubeX = 0     
+    tubeY = DSO    
+
+    # Iso-center
+    isoX = geo.isoX
+    isoY = geo.isoY
+
+    reconImg = zeros(nPix,nPix)
```
Comment (JeffFessler) [inline src/bdd_2d.jl:212]:
```suggestion
    reconImg = zeros(T, nPix, nPix)
```

File: src/bdd_2d.jl [hunk 29]
```diff
@@ -0,0 +1,408 @@
...
+        
+        if ((angleCase == 1)||(angleCase == 3))
+            reconImg = reconImg + reconImgTmp 
+        else
+            reconImg = reconImg + reverse(reconImgTmp, dims = 2)'
+        end 
+           
+    end # Projection loop
+    
+    reconImg = reconImg / length(theta)
```
Comment (JeffFessler) [inline src/bdd_2d.jl:347]:
```suggestion
    reconImg /= length(theta) 
```

File: src/bdd_2d.jl [hunk 30]
```diff
@@ -0,0 +1,408 @@
...
+"""
+function integrate1D(p_v::Vector{T}, pixelSize) where {T <: Number}
+
+    n_pixel = length(p_v)
+    
+    P_x = 0
+    Ppj = zeros(T, n_pixel+1)
+    
+    for pj in 1:n_pixel
+
```
Comment (JeffFessler) [inline src/bdd_2d.jl:368]:
```suggestion
```

File: src/bdd_2d.jl [hunk 31]
```diff
@@ -0,0 +1,408 @@
...
+       P_x += p_v[pj] * pixelSize
+
```
Comment (JeffFessler) [inline src/bdd_2d.jl:370]:
```suggestion
```

File: src/bdd_2d.jl [hunk 32]
```diff
@@ -0,0 +1,408 @@
...
+    
+    P_x = 0
+    Ppj = zeros(T, n_pixel+1)
+    
+    for pj in 1:n_pixel
+
+       P_x += p_v[pj] * pixelSize
+       
+       Ppj[pj+1] = P_x
+
```
Comment (JeffFessler) [inline src/bdd_2d.jl:372]:
```suggestion
```

File: src/bdd_2d.jl [hunk 33]
```diff
@@ -0,0 +1,408 @@
...
+    for pj in 1:n_pixel
+
+       P_x += p_v[pj] * pixelSize
+       
+       Ppj[pj+1] = P_x
+       
+    end
+
+    return Ppj
+
```
Comment (JeffFessler) [inline src/bdd_2d.jl:376]:
```suggestion
```

**Review by JeffFessler — APPROVED** [review]
